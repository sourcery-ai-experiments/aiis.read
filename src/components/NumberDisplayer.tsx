import React, { useMemo } from 'react';
import CircularProgress from '@mui/material/CircularProgress';
import BigNumber from 'bignumber.js';

type NumberDisplayerProps = {
  text?: string;
  className?: string;
  /**
   * @default false
   */
  loading?: boolean;
  /**
   * 有很多个零的情况下最后数字的精度
   * @default 4
   */
  longZeroCasePrecision?: number;
};

/**
 * 展示 bignumber 组件
 *
 * eg. 0.0{4}5252
 */
export function NumberDisplayer({
  text = '0',
  className,
  loading = false,
  longZeroCasePrecision = 4,
}: NumberDisplayerProps) {
  // 防止不会使用，或者错误传错类型，有助于开发阶段尽早发现问题
  if (typeof text !== 'string') throw new Error('text should be string');
  // 转成处理过后的字符串形式
  const number = useMemo(
    () => new BigNumber(text).dividedBy(new BigNumber(Math.pow(10, 18))).toFixed(),
    [text]
  );
  const [valueBeforeDot, valueAfterDot] = number.split('.');
  const computedValue = useMemo(() => {
    let value;
    // 整数补2个0，不用算
    if (valueAfterDot == null) {
      value = '00';
      // 如果 >= 1
    } else if (BigNumber(valueBeforeDot).gte(1)) {
      // 小数部分取2位长度
      value = BigNumber(`0.${valueAfterDot}`).toFixed(2).split('.')[1];
      // 如果 < 1 && 小数部分 > 0.0001
    } else if (
      BigNumber(valueBeforeDot).isLessThan(1) &&
      BigNumber(`0.${valueAfterDot}`).gte(0.0001)
    ) {
      // 小数部分取4个长度
      value = BigNumber(`0.${valueAfterDot}`).toFixed(4).split('.')[1];
      // 其他情况给0计数
    } else {
      const lastZeroIndex = getLastZero(valueAfterDot);
      const zeroCount = valueAfterDot.substring(0, lastZeroIndex + 1).length;
      const noneZeroValue = valueAfterDot.substring(lastZeroIndex + 1);
      const [, noneZeroRealValue] = BigNumber(`0.${noneZeroValue}`)
        .toFixed(longZeroCasePrecision)
        .split('.');
      value = `0{${zeroCount}}${noneZeroRealValue}`;
    }
    return value;
  }, [longZeroCasePrecision, valueAfterDot, valueBeforeDot]);
  // 拆分成2部分

  return (
    <span className={className}>
      {loading ? (
        <CircularProgress size={12} sx={{ marginTop: '6px' }} />
      ) : (
        valueBeforeDot + '.' + computedValue
      )}
    </span>
  );
}

// 返回遇到的第一个下一个数字是非零的零的下标
// 00005203 -> 3
function getLastZero(numStr: string) {
  let index = -1;
  for (let n = 0; n < numStr.length; n++) {
    if (numStr[n] === '0') {
      index = n;
    } else {
      break;
    }
  }
  return index;
}
