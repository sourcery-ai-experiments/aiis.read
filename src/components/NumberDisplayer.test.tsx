import { render } from '@testing-library/react';

import { NumberDisplayer } from './NumberDisplayer';

it('bignumber should work well', () => {
  expect(render(<NumberDisplayer text="3368437500000000" />).container).toMatchSnapshot();
});
it('integer should be append two zero to tail', () => {
  expect(render(<NumberDisplayer text="3000000000000000000" />).container).toMatchSnapshot();
});
it('if interer part great than 1, decimal part should have 2 decimals', () => {
  expect(render(<NumberDisplayer text="3145000000000000000" />).container).toMatchSnapshot();
});
it('if integer part less than 1, decimal part should have 4 decimals', () => {
  expect(render(<NumberDisplayer text="145000000000000000" />).container).toMatchSnapshot();
});
it('if decimal part less than 0.0001, decimal part should round into 4 decimals', () => {
  expect(render(<NumberDisplayer text="000000000012455" />).container).toMatchSnapshot();
});
it('if decimal part less than 0.0001, decimal part should round into 3 decimals', () => {
  expect(
    render(<NumberDisplayer text="000000000012455" longZeroCasePrecision={3} />).container
  ).toMatchSnapshot();
});

it('should throw error if text is not string', () => {
  expect(() =>
    render(<NumberDisplayer text={3.14 as unknown as string} />)
  ).toThrowErrorMatchingSnapshot();
});
