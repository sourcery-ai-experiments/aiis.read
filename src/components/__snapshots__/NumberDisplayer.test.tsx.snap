// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bignumber should work well 1`] = `
<div>
  <span>
    0.0034
  </span>
</div>
`;

exports[`if decimal part less than 0.0001, decimal part should round into 3 decimals 1`] = `
<div>
  <span>
    0.0{13}125
  </span>
</div>
`;

exports[`if decimal part less than 0.0001, decimal part should round into 4 decimals 1`] = `
<div>
  <span>
    0.0{13}1246
  </span>
</div>
`;

exports[`if integer part less than 1, decimal part should have 4 decimals 1`] = `
<div>
  <span>
    0.1450
  </span>
</div>
`;

exports[`if interer part great than 1, decimal part should have 2 decimals 1`] = `
<div>
  <span>
    3.15
  </span>
</div>
`;

exports[`integer should be append two zero to tail 1`] = `
<div>
  <span>
    3.00
  </span>
</div>
`;

exports[`should throw error if text is not string 1`] = `"text should be string"`;
